# Import necessary libraries

In [1]:
import pandas as pd

# Download Data from the API

In [2]:
# Download dataset 
# Dataset Description - Summary information and metrics for Airbnb listings in New_York - 2025-10-01
!wget https://data.insideairbnb.com/united-states/ny/new-york-city/2025-10-01/visualisations/listings.csv -O nyc_listings.csv

--2025-11-15 04:27:13--  https://data.insideairbnb.com/united-states/ny/new-york-city/2025-10-01/visualisations/listings.csv
Resolving data.insideairbnb.com (data.insideairbnb.com)... 3.164.206.36, 3.164.206.19, 3.164.206.7, ...
Connecting to data.insideairbnb.com (data.insideairbnb.com)|3.164.206.36|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5688892 (5.4M) [application/csv]
Saving to: ‘nyc_listings.csv’

nyc_listings.csv    100%[===================>]   5.42M  6.30MB/s    in 0.9s    

2025-11-15 04:27:14 (6.30 MB/s) - ‘nyc_listings.csv’ saved [5688892/5688892]



# 1. Data Loading and Initial Exploration

In [4]:
# Load the dataset
df = pd.read_csv('nyc_listings.csv')

In [5]:
# View the first five rows 
df.head()

id                                               name  host_id  \
0  2595  Skylit Studio Oasis | Midtown Manhattan Sanctuary     2845   
1  6848                   Only 2 stops to Manhattan studio    15991   
2  6872  Uptown Sanctuary w/ Private Bath (Month to Month)    16104   
3  6990                            UES Beautiful Blue Room    16800   
4  7064  Amazing location! Wburg. Large, bright & tranquil    17297   

   host_name neighbourhood_group neighbourhood  latitude  longitude  \
0   Jennifer           Manhattan       Midtown  40.75356  -73.98559   
1      Allen            Brooklyn  Williamsburg  40.70935  -73.95342   
2  Kahshanna           Manhattan   East Harlem  40.80107  -73.94255   
3    Cynthia           Manhattan   East Harlem  40.78778  -73.94759   
4     Joelle            Brooklyn  Williamsburg  40.71248  -73.95881   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt  240.0              30                 47  2022-06-21   
1  Entire home/apt   96.0              30                195  2025-06-09   
2     Private room   59.0              30                  1  2022-06-05   
3     Private room   73.0              30                249  2025-05-27   
4     Private room    NaN              30                 13  2022-09-12   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.24                               3               289   
1               0.98                               1               285   
2               0.02                               2                83   
3               1.28                               1               186   
4               0.07                               2                 0   

   number_of_reviews_ltm license  
0                      0     NaN  
1                      3     NaN  
2                      0     NaN  
3                      4     NaN  
4                      0     NaN

In [6]:
# Initial Info (Column types and non-null counts)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36111 entries, 0 to 36110
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              36111 non-null  int64  
 1   name                            36109 non-null  object 
 2   host_id                         36111 non-null  int64  
 3   host_name                       34945 non-null  object 
 4   neighbourhood_group             36111 non-null  object 
 5   neighbourhood                   36111 non-null  object 
 6   latitude                        36111 non-null  float64
 7   longitude                       36111 non-null  float64
 8   room_type                       36111 non-null  object 
 9   price                           21328 non-null  float64
 10  minimum_nights                  36111 non-null  int64  
 11  number_of_reviews               36111 non-null  int64  
 12  last_review                     

In [7]:
# Initial Description (Summary statistics) including non-numeric columns
df.describe(include='all')

id                            name       host_id host_name  \
count   3.611100e+04                           36109  3.611100e+04     34945   
unique           NaN                           34446           NaN      8296   
top              NaN  Water View King Bed Hotel Room           NaN    Eugene   
freq             NaN                              30           NaN       571   
mean    4.667906e+17                             NaN  1.756195e+08       NaN   
std     5.448721e+17                             NaN  1.945624e+08       NaN   
min     2.595000e+03                             NaN  1.678000e+03       NaN   
25%     2.159416e+07                             NaN  1.795864e+07       NaN   
50%     5.095308e+07                             NaN  8.832602e+07       NaN   
75%     9.747695e+17                             NaN  3.107737e+08       NaN   
max     1.521848e+18                             NaN  7.214431e+08       NaN   

       neighbourhood_group       neighbourhood      latitude     longitude  \
count                36111               36111  36111.000000  36111.000000   
unique                   5                 224           NaN           NaN   
top              Manhattan  Bedford-Stuyvesant           NaN           NaN   
freq                 16081                2612           NaN           NaN   
mean                   NaN                 NaN     40.728253    -73.947350   
std                    NaN                 NaN      0.056269      0.054995   
min                    NaN                 NaN     40.500366    -74.251907   
25%                    NaN                 NaN     40.688260    -73.983560   
50%                    NaN                 NaN     40.725960    -73.955000   
75%                    NaN                 NaN     40.762120    -73.928118   
max                    NaN                 NaN     40.912017    -73.711822   

              room_type         price  minimum_nights  number_of_reviews  \
count             36111  21328.000000    36111.000000       36111.000000   
unique                4           NaN             NaN                NaN   
top     Entire home/apt           NaN             NaN                NaN   
freq              19167           NaN             NaN                NaN   
mean                NaN    680.526819       28.610091          27.321232   
std                 NaN   4480.453282       29.166476          70.308684   
min                 NaN     10.000000        1.000000           0.000000   
25%                 NaN     89.000000       30.000000           0.000000   
50%                 NaN    154.000000       30.000000           3.000000   
75%                 NaN    279.000000       30.000000          23.000000   
max                 NaN  50104.000000     1124.000000        3736.000000   

       last_review  reviews_per_month  calculated_host_listings_count  \
count        24923       24923.000000                    36111.000000   
unique        3400                NaN                             NaN   
top     2025-08-31                NaN                             NaN   
freq           323                NaN                             NaN   
mean           NaN           0.821178                       67.455152   
std            NaN           1.916104                      215.678813   
min            NaN           0.010000                        1.000000   
25%            NaN           0.070000                        1.000000   
50%            NaN           0.250000                        2.000000   
75%            NaN           0.900000                        9.000000   
max            NaN         122.760000                     1148.000000   

        availability_365  number_of_reviews_ltm license  
count       36111.000000           36111.000000    5376  
unique               NaN                    NaN    1962  
top                  NaN                    NaN  Exempt  
freq                 NaN                    NaN    3106  
mean          162.188724               3

In [8]:
# Check for missing values
df.isnull().sum()

id                                    0
name                                  2
host_id                               0
host_name                          1166
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                             14783
minimum_nights                        0
number_of_reviews                     0
last_review                       11188
reviews_per_month                 11188
calculated_host_listings_count        0
availability_365                      0
number_of_reviews_ltm                 0
license                           30735
dtype: int64

In [9]:
# Check for duplicated rows
print(f"Number of duplicated rows: {df.duplicated().sum()}")

Number of duplicated rows: 0


In [10]:
df.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                             float64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
license                            object
dtype: object

# 2. Data Cleaning

In [11]:
# Create a copy of df as working dataset to keep the original intact
df_clean = df.copy()

In [13]:
# Convert 'price' fields (e.g., "$2,100.00") to float
df_clean['price'] = (
    df_clean['price'] \
        .astype(float) \
        .replace('[\$,]', '', regex=True)
)

# Verify conversion
print(f"Price fields data type: {df_clean['price'].dtypes}")

Price fields data type: float64


In [15]:
# Parse dates (e.g., last_review) into datetime objects.
df_clean['last_review'] = pd.to_datetime(df_clean['last_review'], errors='coerce')

print(f"Data type for `last_review` column: {df_clean['last_review'].dtypes}")

Data type for `last_review` column: datetime64[ns]


In [16]:
# Handle missing values in critical fields like reviews_per_month, host_name, and neighbourhood_group.
df_clean['reviews_per_month'] = df_clean['reviews_per_month'].fillna(0)
df_clean['host_name'] = df_clean['host_name'].fillna('Unknown')
df_clean['neighbourhood_group'] = df_clean['neighbourhood_group'].fillna('Other')
df_clean['name'] = df_clean['name'].fillna('Unnamed Listing')
df_clean.head()

id                                               name  host_id  \
0  2595  Skylit Studio Oasis | Midtown Manhattan Sanctuary     2845   
1  6848                   Only 2 stops to Manhattan studio    15991   
2  6872  Uptown Sanctuary w/ Private Bath (Month to Month)    16104   
3  6990                            UES Beautiful Blue Room    16800   
4  7064  Amazing location! Wburg. Large, bright & tranquil    17297   

   host_name neighbourhood_group neighbourhood  latitude  longitude  \
0   Jennifer           Manhattan       Midtown  40.75356  -73.98559   
1      Allen            Brooklyn  Williamsburg  40.70935  -73.95342   
2  Kahshanna           Manhattan   East Harlem  40.80107  -73.94255   
3    Cynthia           Manhattan   East Harlem  40.78778  -73.94759   
4     Joelle            Brooklyn  Williamsburg  40.71248  -73.95881   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt  240.0              30                 47  2022-06-21   
1  Entire home/apt   96.0              30                195  2025-06-09   
2     Private room   59.0              30                  1  2022-06-05   
3     Private room   73.0              30                249  2025-05-27   
4     Private room    NaN              30                 13  2022-09-12   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.24                               3               289   
1               0.98                               1               285   
2               0.02                               2                83   
3               1.28                               1               186   
4               0.07                               2                 0   

   number_of_reviews_ltm license  
0                      0     NaN  
1                      3     NaN  
2                      0     NaN  
3                      4     NaN  
4                      0     NaN

In [17]:
# Remove irrelevant rows (e.g., listings with zero availability or price).
initial_rows = len(df_clean)

df.dropna(subset=['price'], inplace=True)

df_clean = df_clean[
    (df_clean['price'] > 0) &
    (df_clean['availability_365'] > 0)
]

rows_removed = initial_rows - len(df_clean)

print(f'Initial rows: {initial_rows}')
print(f'Final rows after cleaning: {len(df)}')
print(f"Rows dropped (NaN price, price=0, or availability=0): {rows_removed}")

Initial rows: 36111
Final rows after cleaning: 21328
Rows dropped (NaN price, price=0, or availability=0): 14954


# 3. Data Enrichment

In [18]:
# Create a price_per_booking column using price and minimum_nights.
df_clean['price_per_booking'] = (
    df_clean['price'] / df_clean['minimum_nights']
    ).round(2)

df_clean.head()

id                                               name  host_id  \
0  2595  Skylit Studio Oasis | Midtown Manhattan Sanctuary     2845   
1  6848                   Only 2 stops to Manhattan studio    15991   
2  6872  Uptown Sanctuary w/ Private Bath (Month to Month)    16104   
3  6990                            UES Beautiful Blue Room    16800   
7  8490   Maison des Sirenes1,bohemian, luminous apartment    25183   

   host_name neighbourhood_group       neighbourhood   latitude  longitude  \
0   Jennifer           Manhattan             Midtown  40.753560 -73.985590   
1      Allen            Brooklyn        Williamsburg  40.709350 -73.953420   
2  Kahshanna           Manhattan         East Harlem  40.801070 -73.942550   
3    Cynthia           Manhattan         East Harlem  40.787780 -73.947590   
7   Nathalie            Brooklyn  Bedford-Stuyvesant  40.684556 -73.939634   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt  240.0              30                 47  2022-06-21   
1  Entire home/apt   96.0              30                195  2025-06-09   
2     Private room   59.0              30                  1  2022-06-05   
3     Private room   73.0              30                249  2025-05-27   
7  Entire home/apt  170.0              30                189  2023-10-16   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.24                               3               289   
1               0.98                               1               285   
2               0.02                               2                83   
3               1.28                               1               186   
7               0.96                               2               275   

   number_of_reviews_ltm license  price_per_booking  
0                      0     NaN               8.00  
1                      3     NaN               3.20  
2                      0     NaN               1.97  
3                      4     NaN               2.43  
7                      0     NaN               5.67

In [19]:
#- Bucket availability into categories:
# - Full-time (availability > 300)
#- Part-time (100–300)
#- Rare (<100)

def categorize(availability_365):
    if availability_365 > 300:
        return 'Full-time'
    elif 100 <= availability_365 <= 300:
        return 'Part_time'
    else:
        return 'Rare '

df_clean['availability_bucket'] = df_clean['availability_365'].apply(categorize)
df_clean.head()

id                                               name  host_id  \
0  2595  Skylit Studio Oasis | Midtown Manhattan Sanctuary     2845   
1  6848                   Only 2 stops to Manhattan studio    15991   
2  6872  Uptown Sanctuary w/ Private Bath (Month to Month)    16104   
3  6990                            UES Beautiful Blue Room    16800   
7  8490   Maison des Sirenes1,bohemian, luminous apartment    25183   

   host_name neighbourhood_group       neighbourhood   latitude  longitude  \
0   Jennifer           Manhattan             Midtown  40.753560 -73.985590   
1      Allen            Brooklyn        Williamsburg  40.709350 -73.953420   
2  Kahshanna           Manhattan         East Harlem  40.801070 -73.942550   
3    Cynthia           Manhattan         East Harlem  40.787780 -73.947590   
7   Nathalie            Brooklyn  Bedford-Stuyvesant  40.684556 -73.939634   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt  240.0              30                 47  2022-06-21   
1  Entire home/apt   96.0              30                195  2025-06-09   
2     Private room   59.0              30                  1  2022-06-05   
3     Private room   73.0              30                249  2025-05-27   
7  Entire home/apt  170.0              30                189  2023-10-16   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.24                               3               289   
1               0.98                               1               285   
2               0.02                               2                83   
3               1.28                               1               186   
7               0.96                               2               275   

   number_of_reviews_ltm license  price_per_booking availability_bucket  
0                      0     NaN               8.00           Part_time  
1                      3     NaN               3.20           Part_time  
2                      0     NaN               1.97               Rare   
3                      4     NaN               2.43           Part_time  
7                      0     NaN               5.67           Part_time

In [20]:
import os
os.makedirs("data", exist_ok=True)
df_clean.to_csv("data/nyc_cleaned.csv", index=False)

# 4. Data Analysis Using Pandas

#### i. What are the top 10 most expensive neighborhoods by average price?

In [21]:
top_neighborhood = df_clean.groupby('neighbourhood') \
    .agg(avg_price = ('price', 'mean'), count=('price', 'count')) \
    .round(2) \
    .sort_values(by='avg_price', ascending=False) \
    .head(10) \
    .reset_index()

print("Top 10 most expensive neighborhoods by average_price: \n")
top_neighborhood

#print(top_neighborhood.to_markdown(index=False))

Top 10 most expensive neighborhoods by average_price: 



neighbourhood  avg_price  count
0              SoHo    3418.54    134
1  Long Island City    2751.11    254
2       Murray Hill    2359.19    324
3   Lower East Side    2213.08    323
4           Midtown    2168.86   1284
5   Upper West Side    1894.17    827
6            Nolita    1796.10     96
7         Chinatown    1684.53    149
8    Hell's Kitchen    1577.62    966
9   Upper East Side    1145.28    924

#### ii. What’s the average availability and price by room type?

In [22]:
room_type_stats = df_clean.groupby('room_type', as_index=False) \
    .agg(avg_price = ('price', 'mean'), avg_availability = ('availability_365', 'mean')) \
    .round(2) 

print("Average availability and price by room type: \n")
room_type_stats

#print(room_type_stats.to_markdown(index=False))

Average availability and price by room type: 



room_type  avg_price  avg_availability
0  Entire home/apt     307.16            254.49
1       Hotel room   36431.21            293.67
2     Private room     231.42            249.53
3      Shared room     141.04            282.03

#### iii. Which host has the most listings?

In [23]:
most_listings = df_clean.groupby(['host_id', 'host_name'], as_index=False) \
    .agg(qty = ('id',"count")) \
    .sort_values(by="qty", ascending=False) \
    .head(1)

print('Host with the most listings: \n')
most_listings

#print(most_listings.to_markdown(index=False))

Host with the most listings: 



host_id host_name   qty
5198  107434423   Unknown  1019

#### iv. How does average price vary across different boroughs or districts?

In [25]:
borough_price = df_clean.groupby('neighbourhood_group') \
    .agg(avg_price = ("price", "mean"), qty = ("price", "count")) \
    .sort_values(by="avg_price", ascending=False) \
    .reset_index()

print("Average price by borough: \n")
borough_price

#print(borough_price.to_markdown(index=False))

Average price by borough: 



neighbourhood_group    avg_price   qty
0           Manhattan  1182.954830  9276
1              Queens   334.952727  3300
2            Brooklyn   247.781598  7445
3               Bronx   130.626667   825
4       Staten Island   125.556270   311

#### v. How many listings have never been reviewed?

In [26]:
never_reviewed = df_clean[df_clean['number_of_reviews'] == 0] \
    .shape[0]# count number of rows

print(f"The number of listings that have never been reviewed: {never_reviewed}")

The number of listings that have never been reviewed: 6343


# Key Insights

1. **Manhattan borough** is the most premium area for Airbnb listings with an average price of **$1,182.95.**

2. There is a **High Rate of Unreviewed Listings** with approximately 30% (6,343 listings) of the active listings having no reviews. This suggests that a large number of listings are potentially new, non-operational, or have a very low booking rate.

3. The SoHo neighborhood, with an average price of **$3,418.54**, is an extreme outlier among neighborhoods, which aligns with the overall high average price in Manhattan.

4. The host with the last largest portfolio has **1,019**m active listings.

5. Price is strongly correlated with both the type of accommodation and its location

#### Checking!!!